# 5. Callbacks and Promises:

Callbacks and Promises are both mechanisms for handling asynchronous operations in JavaScript.

#### 5.1. What are callbacks and how do they work?

* **What are callbacks:** This are functions that are passed as arguments to other functions and are called when the asynchronous operation is completed. 

* **How do they work:** Callback functions are commonly used in JavaScript to handle asynchronous tasks such as fetching data from an API, processing files, or performing network requests.

> Suppose you want to fetch data from an API using the fetch() method. You would pass a callback function to the then() method, which would be executed when the data is retrieved from the API.

```Javascript
fetch('https://api.example.com/data')
  .then(response => response.json())
  .then(data => console.log(data))
  .catch(error => console.error(error));

```


#### 5.2. Examples of using callbacks to handle asynchronous events.

> In this example, a callback function is registered as an event handler for a button click event. When the button is clicked, the callback function is executed.

```Javascript
const button = document.getElementById("myButton");

button.addEventListener("click", function(event) {
  console.log("Button clicked!");
});
```

#### 5.3. What are promises and how do they simplify asynchronous code?

* **What are promises:** Are objects that represent the eventual completion (or failure) of an asynchronous operation and allow you to chain multiple asynchronous operations together.

> A Promise can be in one of three states: `pending`, `fulfilled`, or `rejected`.

When a Promse is fulfilled or rejected, it returns a value or an error that can be handled with the `.then()` and `.catch()` methods, respectively.

The same data-fetching operation using Promises would look like this:

```Javascript
fetch('https://api.example.com/data')
  .then(response => {
    if (!response.ok) {
      throw new Error('Network response was not ok');
    }
    return response.json();
  })
  .then(data => console.log(data))
  .catch(error => console.error(error));
```
